go/libkb/pvl_interp.go

```diff
@@ -0,0 +1,825 @@
+		return state, NewProofError(keybase1.ProofStatus_CONTENT_FAILURE,
+			"Could not get html for selection: %v", err)
+	}
+
+	state.ActiveString = res
+	return state, nil
+}
+
+func pvlStepTransformURL(g GlobalContextLite, ins *jsonw.Wrapper, state PvlScriptState) (PvlScriptState, ProofError) {
+	sourceTemplate, err := ins.AtKey(PvlTransformURL).GetString()
+	if err != nil {
+		return state, NewProofError(keybase1.ProofStatus_INVALID_PVL,
+			"Could not get source pattern %v", ins)
+	}
+	destTemplate, err := ins.AtKey("to").GetString()
+	if err != nil {
+		return state, NewProofError(keybase1.ProofStatus_INVALID_PVL,
+			"Could not get dest pattern %v", ins)
+	}
+
+	re, perr := pvlInterpretRegex(sourceTemplate, state.Vars)
+	if perr != nil {
+		return state, perr
+	}
+
+	match := re.FindStringSubmatch(state.FetchURL)
+	if len(match) < 1 {
+		g.GetLog().Debug("PVL regex transform did not match: %v %v", re, state.FetchURL)
+		return state, NewProofError(keybase1.ProofStatus_CONTENT_FAILURE,
+			"Regex transform did not match: %v", re)
+	}
+
+	newURL, err := pvlSubstitute(destTemplate, state.Vars, match)
+	if err != nil {
+		g.GetLog().Debug("PVL regex transform did not substitute: %v %v", re, state.FetchURL)
+		return state, NewProofError(keybase1.ProofStatus_BAD_API_URL,
+			"Regex transform did not substitute: %v %v", re, err)
+	}
+
+	state.FetchURL = newURL
+	state.ActiveString = newURL
+	return state, nil
+}
+
+// Run a PVL CSS selector.
+// selectors is a list like [ "div .foo", 0, ".bar"] ].
+// Each string runs a selector, each integer runs a Eq.
+func pvlRunCSSSelectorInner(g GlobalContextLite, html *goquery.Selection, selectors *jsonw.Wrapper) (*goquery.Selection, ProofError) {
+	nselectors, err := selectors.Len()
+	if err != nil {
+		return nil, NewProofError(keybase1.ProofStatus_INVALID_PVL,
+			"Could not get length of selector list")
+	}
+	if nselectors < 1 {
+		return nil, NewProofError(keybase1.ProofStatus_INVALID_PVL,
+			"CSS selectors array must not be empty")
+	}
+
+	var selection *goquery.Selection
+	selection = html
+
+	for i := 0; i < nselectors; i++ {
+		selector := selectors.AtIndex(i)
+
+		selectorIndex, err := selector.GetInt()
+		selectorIsIndex := err == nil
+		selectorString, err := selector.GetString()
+		selectorIsString := err == nil && !selectorIsIndex
+
+		switch {
+		case selectorIsIndex:
+			selection = selection.Eq(selectorIndex)
+		case selectorIsString:
+			selection = selection.Find(selectorString)
+		default:
+			return nil, NewProofError(keybase1.ProofStatus_INVALID_PVL,
+				"Selector entry string or int %v", selector)
+		}
+	}
+
+	return selection, nil
+}
+
+// Most failures here log instead of returning an error. If an error occurs, ([], nil) will be returned.
+// This is because a selector may descend into many subtrees and fail in all but one.
+func pvlRunSelectorJSONInner(g GlobalContextLite, object *jsonw.Wrapper, selectors []*jsonw.Wrapper) ([]string, ProofError) {
+	if len(selectors) == 0 {
```

Comment:
This is the "we've recursively reached the end of the selectors list and should now return an actual value" branch, right? Good place for a comment.